Spring-Boot/Spring-Boot-Academy/src/test/java/tw/waterball/judgegirl/springboot/academy/controllers/ExamControllerTest.java

```diff
@@ -777,6 +781,19 @@ void givenExamWithQuestions_ABC_whenReorderQuestionsWith_213_thenQuestionsOrderS
         assertEquals(3, C.getQuestionOrder());
     }
 
+    @Test
+    void WhenStudentAGetOnGoingExamWithNonWhitelistIpAddress_ThenShouldRespondNotFound() throws Exception {
+        var exam = createExamAndGet(ONGOING_DURATION, "exam");
+        addExaminee(exam.id, STUDENT_A_ID);
+        String nonWhitelistIpAddress = "31.63.127.255";
+        assertFalse(exam.whitelist.contains(nonWhitelistIpAddress));
+
+        mockMvc.perform(withStudentToken(STUDENT_A_ID,
+                        get("/api/exams/{examId}", exam.id))
+                        .with(remoteAddress(nonWhitelistIpAddress)))
+                .andExpect(status().isNotFound());
+    }
```

Comment:
1. Rename to `GivenStudentAJoinAnOnGoingExam_WhenGetExamWithIpAddressNotInWhitelist_ShouldRespondNotFound`.
2. You should have another test about getting an ongoing exam with the IpAddressInTheWhitelist. This is the most valuable user path, so we should protect it with the test.